### PR TITLE
feat: add support for eslint-vue

### PR DESCRIPTION
**Pull request by HerringtonDarkholme**
I choose to support only vanilla eslint-plugin-vue configuration. If users want to use their own configuration, they can simply use `eslint` plugin in their editors.

@mysticatea  Do you have any suggestion?

### CODE REVIEW

File: server/src/modes/template/services/htmlValidation.ts [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+import { configs } from 'eslint-plugin-vue';
+import { TextDocument, Diagnostic, Range, DiagnosticSeverity } from 'vscode-languageserver-types';
+
+export function doValidation(document: TextDocument, engine: CLIEngine): Diagnostic[] {
+  const text = document.getText()
+    .replace(/^\s{10}/, '<template>')
+    .replace(/\s{11}$/, '</template>');
+  const report = engine.executeOnText(text, document.uri);
+  return report.results[0].messages.map(ret => ({
+    range: Range.create(ret.line - 1, ret.column - 1, ret.endLine - 1, ret.endColumn - 1),
```
Comment (mysticatea) [inline server/src/modes/template/services/htmlValidation.ts]: I have notes here:

- `endLine` and `endColumn` may not exist. (see also: https://github.com/Microsoft/vscode-eslint/commit/1d35a9955f6db5b54a27d2e33327a36b83fe967d)
- `line` and `column` may be a negative integer. (see also: https://github.com/Microsoft/vscode-eslint/commit/a552bf12d5aaf8ff28d740e31ccc9ff0bae97007)

Comment (HerringtonDarkholme) [inline server/src/modes/template/services/htmlValidation.ts]: Thanks for pointing this out! Fixed.

**Comment by octref** [comment]
Looks great 👍  I'll update docs at https://octref.github.io/vetur/linting-error.html later.

**Comment by pdanpdan** [comment]
How can I disable this plugin, or configure the rules?

**Comment by pdanpdan** [comment]
Thank you.

**Comment by octref** [comment]
`vetur.validation.vue-html` don't think we have it yet 😅 ...Right?

I'll do it later tonight.

**Comment by rhyek** [comment]
Please. I'm getting loads of errors from `require-v-for-key` that I don't care about.

**Comment by octref** [comment]
`vetur.validation.template option` will be available in the next version.

**Comment by rhyek** [comment]
What is my best option if I want to keep validation but configure it (disable a few checks)? Would it be to disable `vetur.validation.template` once it's possible, install **eslint** and **eslint-plugin-vue** separately and have my own `.eslintrc.js`?.

Thanks.

**Comment by octref** [comment]
Yes, you can disable `vetur.validation.template`, install `eslint` and `eslint-plugin-vue@beta` as your devDependencies and have a `.eslintrc` to fine-tune your linting rules.

**Comment by ErickPetru** [comment]
Guys, I think the documentation is not clear enough about how to customize the linting behavior.

1. I installed `eslint` and `eslint-plugin-vue@beta` locally as `devDependencies`.
2. Then I put in my `.eslintrc.js`:

```
// ...
rules: {
  'vue/html-no-self-closing': 'off'
}
```

3. I keep receiving the error about self-closing tags. Nuxt recommends it and I'd like to stay this way:

```
<template>
  <nuxt />
</template>
```

Of course, if I disable `vetur.validation.template` all the template validations stops... Not what I want.

**Comment by rhyek** [comment]
I just tried setting this up and @ErickPetru is right. There appears to be no way, currently, to configure `<template></template>` linting.

**Comment by octref** [comment]
@rhyek and @ErickPetru 

I just updated `veturpack` https://github.com/octref/veturpack/commit/f0e4c4e89b9edb41e7283c2be56f8c67a780464a

so now it uses `eslint-plugin-vue@beta` and disables template check.
If you have ESLint plugin for VSCode, cloning the repo + `npm i` should make it work. Try editing the `.eslintrc` in the repo.

**Comment by ErickPetru** [comment]
@octref I think we're having some confusion here. I'm talking about lint in **real time**, when VS Code underlines in red the errors while we're working.

I cloned the repo [octref/veturpack](https://github.com/octref/veturpack/blob/master/.eslintrc), but it isn't working too. As you can see, with `vue/html-no-self-closing` set to `error`, we should have `<counter />` underlined in red:

eslint

The `.eslintrc` settings are used only if I manually run `eslint --ext .js,.vue --ignore-path .gitignore .`. 
And if `vetur.validation.template` is set to `true`, the real time linting happens but `.eslintrc` is ignored.

**Comment by HerringtonDarkholme** [comment]
> Install ESLint plugin for the best linting experience. Vetur's template linting is only for quick start and does not support rule configuration.

https://vuejs.github.io/vetur/linting-error.html

Please read the doc

**Comment by ErickPetru** [comment]
Of course, from the doc:

> Vetur automatically uses `eslint-plugin-vue@beta` for linting `<template>`.
> 
> To turn it off, set `vetur.validation.template: false`.
> 
> To configure linting rules, install `eslint` and `eslint-plugin-vue@beta` locally as `devDependencies`, and the ESLint plugin should work.

As we already said, following the doc doesn't let configure linting rules for `<template>`, since `vetur.validation.template: false` fully disables linting inside template (as the picture I posted is showing). What the doc says is invalid, since we can't configure linting rules for `<template>`.

**Comment by octref** [comment]
err

**Comment by ErickPetru** [comment]
@octref I was doing exactly the same thing, without success... Finally I discovered that the problem is with the ESLint extension settings!

I changed my `eslint.validate` inside `settings.json` like explainded [here](https://vuejs.github.io/vetur/setup.html) and I tought it would work immediatally. But its needed to restart VS Code after changing `settings.json`!

Something so simple... Thanks for all the help. Maybe the docs should reforce it, what do you think?

**Comment by octref** [comment]
You can also do Cmd + P -> Reload Window. I think the current doc is fine. If something in VSCode doesn't work, like Linting / IntelliSense, I'd always first try reload project before reporting errors.

**Comment by rhyek** [comment]
Still not working for me.

vetur

**Comment by octref** [comment]
@rhyek You have custom parserOptions for ESLint?

**Comment by rhyek** [comment]
I just removed that section and it is still not working after reloading the window. Not sure why I even had it.

**Comment by octref** [comment]
Does ESLint work for you in Veturpack (https://github.com/octref/veturpack), or it only breaks for your project?

**Comment by rhyek** [comment]
Hi. I'm at work and I'll have to try that later. Something **very** odd I just noticed. Toggling `vue/require-v-for-key` actually works while `vue/html-no-self-closing` doesn't at all.

**Comment by rhyek** [comment]
Actually, I just figured out that the issue is only present in **some** vue files while others work perfectly. I will create an issue over on the `eslint-plugin-vue` repo, but just FYI:

- [productos.vue](https://gist.github.com/rhyek/f3fe31be82ad539c4264204025f933c2): eslint does *not* work.
- [atributos.vue](https://gist.github.com/rhyek/c2703cedd7641589e3942abd1e7cd789): eslint works perfectly.

**Pull request merged**